sys/clif/clif.c

```diff
@@ -0,0 +1,369 @@
+                                            param);
+        if (param_size < 0) {
+            break;
+        }
+        pos_in_link += param_size;
+        link->params_len++;
+    }
+
+    return pos - buf;
+}
+
+ssize_t clif_encode_link(const clif_t *link, char *buf, size_t maxlen)
+{
+    assert(link);
+    size_t pos = 0;
+    ssize_t res = 0;
+
+    res = clif_add_target(link->target, buf, maxlen);
+    if (res < 0) {
+        return res;
+    }
+    pos += res;
+
+    for (unsigned i = 0; i < link->params_len; i++) {
+        res = clif_add_param(&link->params[i], buf ? &buf[pos] : NULL,
+                             maxlen - pos);
+        if (res <= 0) {
+            return res;
+        }
+        pos += res;
+    }
+    return pos;
+}
+
+ssize_t clif_add_target(const char *target, char *buf, size_t maxlen)
+{
+    assert(target);
+
+    size_t pos = 0;
+    size_t target_len = strlen(target);
+
+    if (!buf) {
+        return target_len + 2; /* size after adding '<' and '>' */
+    }
+
+    if ((target_len + 2) > maxlen) {
+        return CLIF_NO_SPACE;
+    }
+
+    buf[pos++] = LF_PATH_BEGIN_C;
+
+    memcpy(&buf[pos], target, target_len);
+    pos += target_len;
+
+    buf[pos++] = LF_PATH_END_C;
+
+    return pos;
+}
+
+ssize_t clif_add_link_separator(char *buf, size_t maxlen)
+{
+    if (!buf) {
+        return 1;
+    }
+
+    if (maxlen < 1) {
+        return CLIF_NO_SPACE;
+    }
+
+    *buf = LF_LINK_SEPARATOR_C;
+    return 1;
+}
+
+ssize_t clif_add_param(clif_param_t *param, char *buf, size_t maxlen)
+{
+    assert(param && param->key);
+
+    /* if no length given, calculate it */
+    if (!param->key_len) {
+        param->key_len = strlen(param->key);
+    }
+
+    /* count param name size and separator ';' */
+    size_t req_space = param->key_len + 1;
+    size_t pos = 0;
+    int quoted = strcmp(param->key, LF_PARAM_SIZE) ? 1 : 0;
+
+    if (param->value) {
+        if (!param->value_len) {
+            param->value_len = strlen(param->value);
+        }
+        /* count also '=' */
+        req_space += param->value_len +  1;
+    }
+
+    if (quoted) {
+        req_space += 2;
+    }
+
+    if (!buf) {
+        return req_space;
+    }
+
+    if (req_space > maxlen) {
+        return CLIF_NO_SPACE;
+    }
+
+    /* add parameter separator ';' */
+    buf[pos++] = LF_PARAM_SEPARATOR_C;
+
+    /* add parameter name */
+    memcpy(&buf[pos], param->key, param->key_len);
+    pos += param->key_len;
+
+    /* add parameter value if defined */
+    if (param->value) {
+        buf[pos++] = LF_PARAM_ATTR_SEPARATOR_C;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+
+        memcpy(&buf[pos], param->value, param->value_len);
+        pos += param->value_len;
+
+        if (quoted) {
+            buf[pos++] = '"';
+        }
+    }
+
+    return pos;
+}
+
+ssize_t clif_get_link(const char *input, size_t input_len, char **output)
```

Comment:
6690 says that the value for a title or extension attribute is a quoted-string, which may include a comma as defined in 2616. See sec 2 intro. Also 2616 sec 2.2 (now 7230 sec 3.2.6) says that a quoted string also may include an escaped quotation mark.

Reply:
So, it looks like these decisions on where one link ends and another begins need to be made on a more localized level, like within clif_get_param(). In this case, I don't see the how clif_get_link() can be useful.

Reply:
You're right, I will change this.

Reply:
Nicely done! I think you've hit the right level of detail. clif_get_attr() looks good.